x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_installation/perform_rule_installation_handler.ts

```diff
@@ -0,0 +1,145 @@
+    const currentRuleVersionsMap = new Map(
+      currentRuleVersions.map((version) => [version.rule_id, version])
+    );
+
+    const allInstallableRules = allLatestVersions.filter((latestVersion) => {
+      const currentVersion = currentRuleVersionsMap.get(latestVersion.rule_id);
+      return !currentVersion;
+    });
+
+    const ruleInstallQueue: Array<{
+      rule_id: RuleSignatureId;
+      version: RuleVersion;
+    }> = [];
+    const ruleErrors = [];
+    const installedRules = [];
+    const skippedRules: SkippedRuleInstall[] = [];
+
+    // Perform all the checks we can before we start the upgrade process
+    if (mode === 'SPECIFIC_RULES') {
+      const installableRuleIds = new Set(allInstallableRules.map((rule) => rule.rule_id));
+      request.body.rules.forEach((rule) => {
+        // Check that the requested rule is not installed yet
+        if (currentRuleVersionsMap.has(rule.rule_id)) {
+          skippedRules.push({
+            rule_id: rule.rule_id,
+            reason: SkipRuleInstallReason.ALREADY_INSTALLED,
+          });
+          return;
+        }
+
+        // Check that the requested rule is installable
+        if (!installableRuleIds.has(rule.rule_id)) {
+          ruleErrors.push({
+            error: new Error(
+              `Rule with ID "${rule.rule_id}" and version "${rule.version}" not found`
+            ),
+            item: rule,
+          });
+          return;
+        }
+
+        ruleInstallQueue.push(rule);
+      });
+    } else if (mode === 'ALL_RULES') {
+      ruleInstallQueue.push(...allInstallableRules);
+    }
+
+    const BATCH_SIZE = 100;
+    while (ruleInstallQueue.length > 0) {
+      const rulesToInstall = ruleInstallQueue.splice(0, BATCH_SIZE);
+      const ruleAssets = await ruleAssetsClient.fetchAssetsByVersion(rulesToInstall);
+
+      const { results, errors } = await createPrebuiltRules(detectionRulesClient, ruleAssets);
+      installedRules.push(...results);
+      ruleErrors.push(...errors);
+    }
+
+    const { error: timelineInstallationError } = await performTimelinesInstallation(
+      ctx.securitySolution
+    );
+
+    const allErrors = aggregatePrebuiltRuleErrors(ruleErrors);
+    if (timelineInstallationError) {
+      allErrors.push({
+        message: timelineInstallationError,
+        rules: [],
+      });
+    }
+
+    const body: PerformRuleInstallationResponseBody = {
+      summary: {
+        total: installedRules.length + skippedRules.length + ruleErrors.length,
+        succeeded: installedRules.length,
+        skipped: skippedRules.length,
+        failed: ruleErrors.length,
+      },
+      results: {
+        created: installedRules.map(({ result }) => result),
+        skipped: skippedRules,
+      },
+      errors: allErrors,
+    };
+
+    return response.ok({ body });
+  } catch (err) {
+    const error = transformError(err);
+    return siemResponse.error({
+      body: error.message,
+      statusCode: error.statusCode,
+    });
+  }
```

Comment:
nit: Catch handler is usually identical for multiple API endpoint handlers. Maybe we could use a hight-order function with that standard logic to even further simplify handlers.